src/NServiceBus.Core/Sagas/ActiveSagaInstance.cs

```diff
@@ -103,51 +114,35 @@ internal void ValidateChanges()
             {
                 ValidateAllCorrelationPropertiesHaveValues();
             }
-            else
-            {
-                ValidateCorrelationPropertiesNotModified();
-            }
+
+            ValidateCorrelationPropertiesNotModified();
         }
 
         void ValidateAllCorrelationPropertiesHaveValues()
         {
-            var properties = Instance.Entity.GetType().GetProperties();
-
-            foreach (var correlatedProperty in Metadata.CorrelationProperties)
+              foreach (var correlationProperty in correlationProperties)
             {
-                var propertyInfo = properties.Single(p => p.Name == correlatedProperty.Name);
-
-                var value = propertyInfo.GetValue(Instance.Entity);
-                var defaultValue = GetDefault(propertyInfo.PropertyType);
+                var defaultValue = GetDefault(correlationProperty.PropertyInfo.PropertyType);
 
-                if (value.Equals(defaultValue))
+                if (correlationProperty.Value.Equals(defaultValue))
                 {
                     throw new Exception(
-                        $@"We detected that the correlated property '{correlatedProperty.Name}' on saga '{Metadata.SagaType.Name}' does not have a value'. 
+                        $@"We detected that the correlated property '{correlationProperty.PropertyInfo.Name}' on saga '{Metadata.SagaType.Name}' does not have a value'. 
 All correlated properties must have a non null or empty value assigned to them when a new saga instance is created.");
                 }
-
-                correlatedPropertyValues[correlatedProperty.Name] = value;
             }
         }
 
         void ValidateCorrelationPropertiesNotModified()
         {
-            if (!correlatedPropertyValues.Any())
+            foreach (var correlationProperty in correlationProperties.Where(cp=>cp.HasExistingValue))
             {
-                return;
-            }
-
-            var properties = Instance.Entity.GetType().GetProperties();
-            foreach (var existingPropertyValue in correlatedPropertyValues)
-            {
-                var propertyInfo = properties.Single(p => p.Name == existingPropertyValue.Key);
-                var currentValue = propertyInfo.GetValue(Instance.Entity);
+                var currentValue = correlationProperty.PropertyInfo.GetValue(Instance.Entity);
 
-                if (!existingPropertyValue.Value.Equals(currentValue))
+                if (correlationProperty.Value.ToString() != currentValue.ToString())
```

Comment:
I this case I choose to always store as a string since just storing the ref means that the value can change underneath (at least that's what I saw with guid's). 

So in essense I store a string "snapshot" so that I can check if the value has been changed. 

Does this make sense? Other ideas?